src/proto/uprotocol/uuri.rs

```diff
@@ -11,38 +11,617 @@
+    ///         ..Default::default()
+    ///     })
+    ///     .into(),
+    ///     resource: Some(UResource {
+    ///         name: "rpc".to_string(),
+    ///         instance: Some("raise".to_string()),
+    ///         ..Default::default()
+    ///     })
+    ///     .into(),
+    ///     authority: None.into(),
+    ///     ..Default::default()
+    /// };
+    ///
+    /// let uri_from = String::try_from(&uri).unwrap();
+    /// assert_eq!("/example.com//rpc.raise", uri_from);
+    /// ````
+    fn try_from(uri: &UUri) -> Result<Self, Self::Error> {
+        if UriValidator::is_empty(uri) {
+            return Err(SerializationError::new("URI is empty"));
+        }
+
+        let mut output = String::default();
+        if let Some(authority) = uri.authority.as_ref() {
+            output.push_str(UUri::build_authority_part_of_uri(authority).as_str());
+        }
+        output.push('/');
+        if let Some(entity) = uri.entity.as_ref() {
+            output.push_str(UUri::build_entity_part_of_uri(entity).as_str());
+        }
+        output.push_str(UUri::build_resource_part_of_uri(uri).as_str());
+
+        // remove trailing slashes
+        Ok(Regex::new(r"/+$")
+            .unwrap()
+            .replace_all(&output, "")
+            .into_owned())
+    }
+}
+
+impl FromStr for UUri {
+    type Err = SerializationError;
+
+    /// Attempts to serialize a `String` into a `UUri`.
+    ///
+    /// # Arguments
+    ///
+    /// * `uri` - The `String` to be converted into a `UUri`.
+    ///
+    /// # Returns
+    ///
+    /// A `Result` containing either the `UUri` representation of the URI or a `SerializationError`.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use std::str::FromStr;
+    /// use up_rust::uprotocol::{UAuthority, UEntity, UResource};
+    /// use up_rust::uprotocol::UUri;
+    ///
+    /// let uri = UUri {
+    ///     entity: Some(UEntity {
+    ///         name: "example.com".to_string(),
+    ///         ..Default::default()
+    ///     })
+    ///     .into(),
+    ///     resource: Some(UResource {
+    ///         name: "rpc".to_string(),
+    ///         instance: Some("raise".to_string()),
+    ///         ..Default::default()
+    ///     })
+    ///     .into(),
+    ///     authority: None.into(),
+    ///     ..Default::default()
+    /// };
+    ///
+    /// let uri_from = UUri::from_str("/example.com//rpc.raise").unwrap();
+    /// assert_eq!(uri, uri_from);
+    /// ````
+    fn from_str(uri: &str) -> Result<Self, Self::Err> {
+        if uri.is_empty() {
+            return Err(SerializationError::new("URI is empty"));
+        }
+
+        let uri = if let Some(index) = uri.find(':') {
+            uri[index + 1..].to_string()
+        } else {
+            uri.replace('\\', "/")
+        };
+        let is_local: bool = !uri.starts_with("//");
+        let uri_parts = Self::pattern_split(&uri, "/");
+
+        if uri_parts.len() < 2 {
```

Comment:
I'm reading through the UUri spec now to parse out what's going on here. May be nice to add a little bit of explanatory comments.

e.g. here could be
```
// If we don't have at a minimum a UEntity and UResource, we fail
```
or something to that effect

WDYT?

Reply:
Or perhaps even better, the error message could be updated to reflect this e.g.
```rust
return Err(SerializationError::new("URI missing UEntity or UResource"));
```

Reply:
Picking the second option :)